File: src/domain/model/canonical_story.py
"""Canonical story — distilled from routa's `canonical-story.ts`.

A canonical story is the structured contract that Backlog produces and Todo
trusts. It is the single source of truth for ``problem_statement``,
``acceptance_criteria``, ``invest`` checks, and dependencies.

We accept the YAML block embedded inside a fenced ```yaml``` region in the
card body. Parsing failures, missing keys, and INVEST FAILs are all
reported as structured ``CanonicalStoryIssue`` items so the UI can render
them as a checklist instead of crashing on the LLM's output.

Per Agent-First: this module only parses + checks **structural** invariants
(field presence, type, INVEST status enum). Whether an AC is "actually
testable" remains a subjective call delegated upstream.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

import yaml

from src.domain.shared_kernel import ValueObject


class InvestStatus(str, Enum):
    PASS = "pass"
    FAIL = "fail"
    WARNING = "warning"


_INVEST_KEYS: tuple[str, ...] = (
    "independent",
    "negotiable",
    "valuable",
    "estimable",
    "small",
    "testable",
)


@dataclass(frozen=True, kw_only=True)
class AcceptanceCriterion(ValueObject):
    id: str
    text: str
    testable: bool


@dataclass(frozen=True, kw_only=True)
class InvestCheck(ValueObject):
    key: str
    status: InvestStatus
    reason: str = ""


@dataclass(frozen=True, kw_only=True)
class CanonicalStory(ValueObject):
    """Parsed canonical story document."""

    version: int
    language: str
    title: str
    problem_statement: str
    user_value: str
    acceptance_criteria: tuple[AcceptanceCriterion, ...]
    constraints_and_affected_areas: tuple[str, ...] = field(default_factory=tuple)
    out_of_scope: tuple[str, ...] = field(default_factory=tuple)
    invest: tuple[InvestCheck, ...] = field(default_factory=tuple)
    depends_on: tuple[str, ...] = field(default_factory=tuple)
    independent_story_check: str = "pass"  # "pass" | "fail"
    unblock_condition: str = ""


@dataclass(frozen=True, kw_only=True)
class CanonicalStoryIssue(ValueObject):
    path: str
    message: str


@dataclass(frozen=True, kw_only=True)
class CanonicalStoryParseResult(ValueObject):
    has_yaml_block: bool
    story: CanonicalStory | None
    issues: tuple[CanonicalStoryIssue, ...] = field(default_factory=tuple)
    raw_yaml: str | None = None

    @property
    def is_valid(self) -> bool:
        return self.story is not None and not self.issues


_YAML_BLOCK_RE = re.compile(r"```yaml\s*\n([\s\S]*?)\n```", re.IGNORECASE)


def parse_canonical_story(card_body: str) -> CanonicalStoryParseResult:
    """Extract the first ```yaml``` block and validate it.

    Returns a result with ``story=None`` on any structural failure; specific
    failure points are captured in ``issues``. Successful parses produce a
    fully-typed ``CanonicalStory``.
    """
    match = _YAML_BLOCK_RE.search(card_body)
    if not match:
        return CanonicalStoryParseResult(
            has_yaml_block=False,
            story=None,
            issues=(
                CanonicalStoryIssue(
                    path="card_body",
                    message="No fenced ```yaml block found.",
                ),
            ),
        )

    raw = match.group(1)
    try:
        loaded = yaml.safe_load(raw)
    except yaml.YAMLError as exc:
        return CanonicalStoryParseResult(
            has_yaml_block=True,
            story=None,
            raw_yaml=raw,
            issues=(
                CanonicalStoryIssue(path="yaml", message=f"YAML parse error: {exc}"),
            ),
        )

    if not isinstance(loaded, dict) or "story" not in loaded:
        return CanonicalStoryParseResult(
            has_yaml_block=True,
            story=None,
            raw_yaml=raw,
            issues=(
                CanonicalStoryIssue(
                    path="yaml.story",
                    message="Top-level `story` key missing.",
                ),
            ),
        )

    story_node = loaded["story"]
    if not isinstance(story_node, dict):
        return CanonicalStoryParseResult(
            has_yaml_block=True,
            story=None,
            raw_yaml=raw,
            issues=(
                CanonicalStoryIssue(
                    path="yaml.story",
                    message="`story` must be a mapping.",
                ),
            ),
        )

    issues: list[CanonicalStoryIssue] = []
    story = _build_story(story_node, issues)
    if issues and story is None:
        return CanonicalStoryParseResult(
            has_yaml_block=True,
            story=None,
            raw_yaml=raw,
            issues=tuple(issues),
        )
    return CanonicalStoryParseResult(
        has_yaml_block=True,
        story=story,
        raw_yaml=raw,
        issues=tuple(issues),
    )


def _build_story(
    node: dict[str, Any], issues: list[CanonicalStoryIssue]
) -> CanonicalStory | None:
    title = _read_str(node, "title", issues)
    problem = _read_str(node, "problem_statement", issues)
    if not title or not problem:
        return None

    acs = _read_acceptance_criteria(node.get("acceptance_criteria"), issues)
    invest = _read_invest(node.get("invest"), issues)

    deps_node = node.get("dependencies_and_sequencing") or {}
    depends_on = _read_str_tuple(deps_node.get("depends_on"))
    independent_check = (
        str(deps_node.get("independent_story_check", "pass")).lower()
        if isinstance(deps_node, dict)
        else "pass"
    )

    return CanonicalStory(
        version=int(node.get("version") or 1),
        language=str(node.get("language") or "en"),
        title=title,
        problem_statement=problem,
        user_value=str(node.get("user_value") or ""),
        acceptance_criteria=acs,
        constraints_and_affected_areas=_read_str_tuple(
            node.get("constraints_and_affected_areas")
        ),
        out_of_scope=_read_str_tuple(node.get("out_of_scope")),
        invest=invest,
        depends_on=depends_on,
        independent_story_check=independent_check,
        unblock_condition=str(deps_node.get("unblock_condition", ""))
        if isinstance(deps_node, dict)
        else "",
    )


def _read_str(node: dict[str, Any], key: str, issues: list[CanonicalStoryIssue]) -> str:
    value = node.get(key)
    if not isinstance(value, str) or not value.strip():
        issues.append(
            CanonicalStoryIssue(
                path=f"story.{key}",
                message=f"`{key}` must be a non-empty string.",
            )
        )
        return ""
    return value.strip()


def _read_str_tuple(value: object) -> tuple[str, ...]:
    if not isinstance(value, list):
        return ()
    return tuple(str(item).strip() for item in value if str(item).strip())


def _read_acceptance_criteria(
    value: object, issues: list[CanonicalStoryIssue]
) -> tuple[AcceptanceCriterion, ...]:
    if not isinstance(value, list) or not value:
        issues.append(
            CanonicalStoryIssue(
                path="story.acceptance_criteria",
                message="acceptance_criteria must be a non-empty list.",
            )
        )
        return ()
    out: list[AcceptanceCriterion] = []
    for idx, raw in enumerate(value):
        if not isinstance(raw, dict):
            issues.append(
                CanonicalStoryIssue(
                    path=f"story.acceptance_criteria[{idx}]",
                    message="Each AC must be a mapping with id/text/testable.",
                )
            )
            continue
        ac_id = str(raw.get("id") or f"AC{idx + 1}").strip()
        text = str(raw.get("text") or "").strip()
        if not text:
            issues.append(
                CanonicalStoryIssue(
                    path=f"story.acceptance_criteria[{idx}].text",
                    message="AC text is required.",
                )
            )
            continue
        out.append(
            AcceptanceCriterion(
                id=ac_id,
                text=text,
                testable=bool(raw.get("testable", False)),
            )
        )
    return tuple(out)


def _read_invest(
    value: object, issues: list[CanonicalStoryIssue]
) -> tuple[InvestCheck, ...]:
    if not isinstance(value, dict):
        issues.append(
            CanonicalStoryIssue(
                path="story.invest",
                message="invest must be a mapping with the six INVEST keys.",
            )
        )
        return ()
    out: list[InvestCheck] = []
    for key in _INVEST_KEYS:
        sub = value.get(key)
        if not isinstance(sub, dict):
            issues.append(
                CanonicalStoryIssue(
                    path=f"story.invest.{key}",
                    message=f"invest.{key} must be a mapping with status + reason.",
                )
            )
            continue
        status_raw = str(sub.get("status", "")).lower()
        try:
            status = InvestStatus(status_raw)
        except ValueError:
            issues.append(
                CanonicalStoryIssue(
                    path=f"story.invest.{key}.status",
                    message=f"invest.{key}.status must be pass/fail/warning.",
                )
            )
            continue
        out.append(
            InvestCheck(
                key=key,
                status=status,
                reason=str(sub.get("reason") or "").strip(),
            )
        )
    return tuple(out)


def invest_blocking_failures(story: CanonicalStory) -> tuple[InvestCheck, ...]:
    """Return INVEST checks that are FAIL (warnings allowed).

    Used by Todo's entry-gate: a story with any FAIL in
    independent/valuable/small/testable should not move forward.
    """
    blocking_keys = {"independent", "valuable", "small", "testable"}
    return tuple(
        c for c in story.invest if c.key in blocking_keys and c.status is InvestStatus.FAIL
    )


__all__ = [
    "AcceptanceCriterion",
    "CanonicalStory",
    "CanonicalStoryIssue",
    "CanonicalStoryParseResult",
    "InvestCheck",
    "InvestStatus",
    "invest_blocking_failures",
    "parse_canonical_story",
]

File: web/src/components/agent/canonicalStory/canonicalStory.ts
/**
 * Canonical Story schema + parser.
 *
 * A "Canonical Story" is a structured YAML block embedded in a markdown
 * fenced code block (```yaml). It captures problem statement, acceptance
 * criteria, INVEST checks, and dependency / sequencing metadata so the UI
 * can render a rich card instead of a wall of text.
 *
 * Distilled from Routa's `src/core/kanban/canonical-story.ts`.
 */

import yaml from 'js-yaml';

export type CanonicalStoryStatus = 'pass' | 'fail' | 'warning';

export interface CanonicalStoryAcceptanceCriterion {
  id: string;
  text: string;
  testable: boolean;
}

export interface CanonicalStoryInvestCheck {
  status: CanonicalStoryStatus;
  reason: string;
}

export interface CanonicalStoryDocument {
  story: {
    version: number;
    language: string;
    title: string;
    problem_statement: string;
    user_value: string;
    acceptance_criteria: CanonicalStoryAcceptanceCriterion[];
    constraints_and_affected_areas: string[];
    dependencies_and_sequencing: {
      independent_story_check: 'pass' | 'fail';
      depends_on: string[];
      unblock_condition: string;
    };
    out_of_scope: string[];
    invest: {
      independent: CanonicalStoryInvestCheck;
      negotiable: CanonicalStoryInvestCheck;
      valuable: CanonicalStoryInvestCheck;
      estimable: CanonicalStoryInvestCheck;
      small: CanonicalStoryInvestCheck;
      testable: CanonicalStoryInvestCheck;
    };
  };
}

export interface CanonicalStoryParseResult {
  story: CanonicalStoryDocument | null;
  issues: string[];
  rawYaml: string;
}

const CANONICAL_STORY_STATUS = new Set<CanonicalStoryStatus>(['pass', 'fail', 'warning']);
const INVEST_KEYS = [
  'independent',
  'negotiable',
  'valuable',
  'estimable',
  'small',
  'testable',
] as const;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function readString(value: unknown, path: string, issues: string[]): string {
  if (typeof value !== 'string' || !value.trim()) {
    issues.push(`${path} must be a non-empty string`);
    return '';
  }
  return value.trim();
}

function readBoolean(value: unknown, path: string, issues: string[]): boolean {
  if (typeof value !== 'boolean') {
    issues.push(`${path} must be a boolean`);
    return false;
  }
  return value;
}

function readStringArray(
  value: unknown,
  path: string,
  issues: string[],
  options?: { optional?: boolean; allowEmpty?: boolean }
): string[] {
  const optional = options?.optional ?? false;
  const allowEmpty = options?.allowEmpty ?? false;
  if ((value === undefined || value === null) && optional) return [];
  if (!Array.isArray(value)) {
    issues.push(`${path} must be an array of strings`);
    return [];
  }
  const items = value
    .map((item, idx) => readString(item, `${path}[${String(idx)}]`, issues))
    .filter(Boolean);
  if (!allowEmpty && items.length === 0) {
    issues.push(`${path} must contain at least one non-empty string`);
  }
  return items;
}

function readAcceptanceCriteria(
  value: unknown,
  path: string,
  issues: string[]
): CanonicalStoryAcceptanceCriterion[] {
  if (!Array.isArray(value)) {
    issues.push(`${path} must be an array`);
    return [];
  }
  const out = value.flatMap((item, idx): CanonicalStoryAcceptanceCriterion[] => {
    if (!isRecord(item)) {
      issues.push(`${path}[${String(idx)}] must be an object`);
      return [];
    }
    return [
      {
        id: readString(item.id, `${path}[${String(idx)}].id`, issues),
        text: readString(item.text, `${path}[${String(idx)}].text`, issues),
        testable: readBoolean(item.testable, `${path}[${String(idx)}].testable`, issues),
      },
    ];
  });
  if (out.length < 2) issues.push(`${path} must contain at least 2 acceptance criteria`);
  return out;
}

function readInvestCheck(
  value: unknown,
  path: string,
  issues: string[]
): CanonicalStoryInvestCheck {
  if (!isRecord(value)) {
    issues.push(`${path} must be an object`);
    return { status: 'fail', reason: '' };
  }
  const statusStr = readString(value.status, `${path}.status`, issues) as CanonicalStoryStatus;
  if (!CANONICAL_STORY_STATUS.has(statusStr)) {
    issues.push(`${path}.status must be one of: pass, fail, warning`);
  }
  return {
    status: CANONICAL_STORY_STATUS.has(statusStr) ? statusStr : 'fail',
    reason: readString(value.reason, `${path}.reason`, issues),
  };
}

/**
 * True if the raw yaml content looks like a canonical-story block — used to
 * cheaply gate parsing without taking the cost of a full yaml load on every
 * fenced code block.
 */
export function looksLikeCanonicalStory(rawYaml: string): boolean {
  return /^\s*story\s*:/m.test(rawYaml);
}

export function parseCanonicalStory(rawYaml: string): CanonicalStoryParseResult {
  let parsed: unknown;
  try {
    parsed = yaml.load(rawYaml);
  } catch (error) {
    const msg = error instanceof Error ? error.message : String(error);
    return { story: null, issues: [`Failed to parse YAML: ${msg}`], rawYaml };
  }

  const issues: string[] = [];
  if (!isRecord(parsed)) {
    return { story: null, issues: ['Canonical story root must be an object'], rawYaml };
  }
  const root = parsed.story;
  if (!isRecord(root)) {
    return { story: null, issues: ['story must be an object'], rawYaml };
  }

  const version = root.version;
  if (typeof version !== 'number' || !Number.isInteger(version) || version < 1) {
    issues.push('story.version must be a positive integer');
  }

  const dependencies = root.dependencies_and_sequencing;
  const invest = root.invest;
  if (!isRecord(dependencies)) issues.push('story.dependencies_and_sequencing must be an object');
  if (!isRecord(invest)) issues.push('story.invest must be an object');

  const story: CanonicalStoryDocument = {
    story: {
      version:
        typeof version === 'number' && Number.isInteger(version) && version > 0 ? version : 1,
      language: readString(root.language, 'story.language', issues),
      title: readString(root.title, 'story.title', issues),
      problem_statement: readString(root.problem_statement, 'story.problem_statement', issues),
      user_value: readString(root.user_value, 'story.user_value', issues),
      acceptance_criteria: readAcceptanceCriteria(
        root.acceptance_criteria,
        'story.acceptance_criteria',
        issues
      ),
      constraints_and_affected_areas: readStringArray(
        root.constraints_and_affected_areas,
        'story.constraints_and_affected_areas',
        issues
      ),
      dependencies_and_sequencing: {
        independent_story_check:
          isRecord(dependencies) &&
          (dependencies.independent_story_check === 'pass' ||
            dependencies.independent_story_check === 'fail')
            ? dependencies.independent_story_check
            : 'fail',
        depends_on: isRecord(dependencies)
          ? readStringArray(
              dependencies.depends_on,
              'story.dependencies_and_sequencing.depends_on',
              issues,
              { optional: true, allowEmpty: true }
            )
          : [],
        unblock_condition: isRecord(dependencies)
          ? readString(
              dependencies.unblock_condition,
              'story.dependencies_and_sequencing.unblock_condition',
              issues
            )
          : '',
      },
      out_of_scope: readStringArray(root.out_of_scope, 'story.out_of_scope', issues, {
        optional: true,
        allowEmpty: true,
      }),
      invest: {
        independent: isRecord(invest)
          ? readInvestCheck(invest.independent, 'story.invest.independent', issues)
          : { status: 'fail', reason: '' },
        negotiable: isRecord(invest)
          ? readInvestCheck(invest.negotiable, 'story.invest.negotiable', issues)
          : { status: 'fail', reason: '' },
        valuable: isRecord(invest)
          ? readInvestCheck(invest.valuable, 'story.invest.valuable', issues)
          : { status: 'fail', reason: '' },
        estimable: isRecord(invest)
          ? readInvestCheck(invest.estimable, 'story.invest.estimable', issues)
          : { status: 'fail', reason: '' },
        small: isRecord(invest)
          ? readInvestCheck(invest.small, 'story.invest.small', issues)
          : { status: 'fail', reason: '' },
        testable: isRecord(invest)
          ? readInvestCheck(invest.testable, 'story.invest.testable', issues)
          : { status: 'fail', reason: '' },
      },
    },
  };

  if (
    !isRecord(dependencies) ||
    (dependencies.independent_story_check !== 'pass' &&
      dependencies.independent_story_check !== 'fail')
  ) {
    issues.push('story.dependencies_and_sequencing.independent_story_check must be pass or fail');
  }
  if (isRecord(invest)) {
    for (const key of INVEST_KEYS) {
      if (!(key in invest)) issues.push(`story.invest.${key} is required`);
    }
  }

  return { story: issues.length === 0 ? story : null, issues, rawYaml };
}

export type CanonicalStoryInvestKey = (typeof INVEST_KEYS)[number];
export const CANONICAL_STORY_INVEST_KEYS = INVEST_KEYS;
